### ISSUE TITLE
Addition & Deletion of NodePool using aks-engine - 0.67.0 & AKS Version 1.20.11

### ISSUE THREAD

**Issue by zohebs341** [Azure/aks-engine#4989]
Hi Team,

I've deployed the AKS cluster using aks-engine on Azure Stack Hub and it's up and running. However, I need to perform some modifications in the cluster. Like I need to delete one node pool and add a new node pool.

my "apimodel.json" file looks like the below:

cat apimodel.json | grep -i profile

    "orchestratorProfile": {
    "masterProfile": {
      "storageProfile": "ManagedDisks",
      "availabilityProfile": "AvailabilitySet",
    "agentPoolProfiles": [
        "availabilityProfile": "AvailabilitySet",
        "storageProfile": "ManagedDisks",
        "availabilityProfile": "AvailabilitySet",
        "storageProfile": "ManagedDisks",

In the portal, I don't see any VMs in the Virtual machine scale sets tab. How can I add a node pool & delete a node pool?

**Comment by zohebs341** [comment]
As I've used the availability set, I need commands/process to add a new node pool & delete the node pool.

**Comment by jadarsie** [comment]
Check this out
https://github.com/Azure/aks-engine/blob/master/docs/topics/addpool.md

**Comment by jadarsie** [comment]
> instead of VMSS I've to use "AvailabilitySet" in the below command right?

Correct, it should contain all the customizations required for ASH (ex `"AcceleratedNetworkingEnabled": false`)

**Comment by zohebs341** [comment]
in apimodel.json, below is for one nodepool. I will use the same values: Except nodepool name, count, vmSize. Hopefully it work with this.

{
        "name": "nodepoolname",
        "count": 1,
        "vmSize": "Standard_D2_v3",
        "osType": "Linux",
        "availabilityProfile": "AvailabilitySet",
        "storageProfile": "ManagedDisks",
        "distro": "aks-ubuntu-18.04",
        "kubernetesConfig": {
          "kubeletConfig": {
            "--address": "0.0.0.0",
            "--anonymous-auth": "false",
            "--authentication-token-webhook": "true",
            "--authorization-mode": "Webhook",
            "--azure-container-registry-config": "/etc/kubernetes/azure.json",
            "--cgroups-per-qos": "true",
            "--client-ca-file": "/etc/kubernetes/certs/ca.crt",
            "--cloud-config": "/etc/kubernetes/azure.json",
            "--cloud-provider": "azure",
            "--cluster-dns": "10.0.0.10",
            "--cluster-domain": "cluster.local",
            "--enforce-node-allocatable": "pods",
            "--event-qps": "0",
            "--eviction-hard": "memory.available<750Mi,nodefs.available<10%,nodefs.inodesFree<5%",
            "--feature-gates": "ExecProbeTimeout=true,RotateKubeletServerCertificate=true",
            "--healthz-port": "10248",
            "--image-gc-high-threshold": "85",
            "--image-gc-low-threshold": "80",
            "--image-pull-progress-deadline": "30m",
            "--keep-terminated-pod-volumes": "false",
            "--kubeconfig": "/var/lib/kubelet/kubeconfig",
            "--max-pods": "110",
            "--network-plugin": "kubenet",
            "--node-status-update-frequency": "1m",
            "--non-masquerade-cidr": "0.0.0.0/0",
            "--pod-infra-container-image": "mcr.microsoft.com/oss/kubernetes/pause:3.4.1",
            "--pod-manifest-path": "/etc/kubernetes/manifests",
            "--pod-max-pids": "-1",
            "--protect-kernel-defaults": "true",
            "--read-only-port": "0",
            "--resolv-conf": "/run/systemd/resolve/resolv.conf",
            "--rotate-certificates": "true",
            "--streaming-connection-idle-timeout": "4h",
            "--tls-cert-file": "/etc/kubernetes/certs/kubeletserver.crt",
            "--tls-cipher-suites": "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_GCM_SHA256",
            "--tls-private-key-file": "/etc/kubernetes/certs/kubeletserver.key",
            "--v": "2",
            "--volume-plugin-dir": "/etc/kubernetes/volumeplugins"
          },
          "cloudProviderBackoffMode": ""
        },
        "acceleratedNetworkingEnabled": false,
        "acceleratedNetworkingEnabledWindows": false,
        "auditDEnabled": false,
        "fqdn": "",
        "preProvisionExtension": null,
        "extensions": [],
        "platformFaultDomainCount": 3,
        "platformUpdateDomainCount": 3,
        "enableVMSSNodePublicIP": false,
        "sysctldConfig": {
          "net.core.message_burst": "80",
          "net.core.message_cost": "40",
          "net.core.somaxconn": "16384",
          "net.ipv4.neigh.default.gc_thresh1": "4096",
          "net.ipv4.neigh.default.gc_thresh2": "8192",
          "net.ipv4.neigh.default.gc_thresh3": "16384",
          "net.ipv4.tcp_max_syn_backlog": "16384",
          "net.ipv4.tcp_retries2": "8"
        },
        "osDiskCachingType": "ReadWrite",
        "dataDiskCachingType": "ReadOnly"
      }

**Comment by zohebs341** [comment]
@jadarsie I've to delete the old node pool and create a new node pool. For deletion of the node pool as well, can you please assist?

I am aware of the drain/cordon of nodes. But completely I wanted to remove one node pool. If I drain/cordon nodes and from the portal If I delete nodes, I am just afraid of apimodel.json file. As the changes should reflect in that as well. So that in future, if I add one more node pool, the apimodel.json should be in sync with current infra/nodes.

**Comment by jadarsie** [comment]
I do not think AKS Engine has a command to delete node pools. I would try these manual steps:

1. drain and cordon nodes
1. delete IaaS
1. kubectl-delete nodes
1. remove nodepool from API Model

The naming convention of a node pool's IaaS resources is `k8s-<nodepool-name>-<cluster-id>-*`. The naming convention for the VMAS is `<nodepool-name>-availabilitySet-<cluster-id>`.

**Comment by zohebs341** [comment]
@jadarsie  Add node pool is failing with the below error.

aks-engine addpool --subscription-id xxxx --resource-group xxxx --location xxxx --api-model apimodel.json --node-pool newnodepool.json --auth-method client_secret --client-id xxxx --client-secret xxxx
WARN[0000] linuxProfile.enableUnattendedUpgrades configuration was not declared, your cluster nodes will be configured to run unattended-upgrade by default
INFO[0000] Writing cloud profile to: /tmp/azurestackcloud.json670337102
Error: failed to load existing container service: failed to get client: Unexpected response from Get Subscription: 404

Even though I've tried using the below command, but still have the same error. 

Command: aks-engine addpool --subscription-id xxxx --resource-group xxxx --location xxxx --api-model apimodel.json --node-pool newnodepool.json

**Comment by zohebs341** [comment]
The command I've used for cluster creation is below one, I am suspecting that client-id & client-secret parameters somehow need to pass.

aks-engine deploy \
--azure-env AzureStackCloud \
--location <for asdk is local> \
--resource-group kube-rg \
--api-model ./kubernetes-azurestack.json \
--output-directory kube-rg \
--client-id xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx \
--client-secret xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx \
--subscription-id xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx

**Comment by zohebs341** [comment]
@jadarsie  Is it mandatory to delete nodepool first and then add new nodepool? Why because, there might be cases where we just need to add new nodepool.

I've deleted old nodepool in my case and trying to add new nodepool. I'm getting few errors which I pasted above. Please assist

**Comment by zohebs341** [comment]
@jadarsie Below command worked for me, for adding the node pool.

aks-engine addpool --subscription-id xxx --azure-env AzureStackCloud --location xxx --resource-group xxx-rg --api-model ./apimodel.json --node-pool ./pool.json --client-id xxx
 --client-secret "xxx"
